### PR TITLE
[tensorflow] [build] Fix TF 2.5 docker image build.

### PR DESCRIPTION
**Pull request by matherit**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Resolves issue with TensorFlow docker image not building due to changes in Nvidia's repo support.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/sanity/test_utility_installation.py [hunk 1]
```diff
@@ -113,12 +114,13 @@ def test_emacs(image):
...
-    if framework not in utility_package_minimum_framework_version or Version(framework_version) < Version(
-        utility_package_minimum_framework_version[framework]) or Version(framework_version) > Version(
-        utility_package_maximum_framework_version[framework]
+    utility_package_framework_version_limit = {
+        "pytorch": SpecifierSet(">=1.7,<1.9"), "tensorflow": SpecifierSet(">=2.4,<2.7,!=2.5.*")
+    }
+
+    if (
+        framework not in utility_package_framework_version_limit
+        or Version(framework_version) not in utility_package_framework_version_limit[framework]
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_utility_installation.py:123]: I have added this change because of build errors in installing these packages through the dockerfile.

File: tensorflow/inference/docker/2.5/py3/cu112/Dockerfile.diy.gpu [hunk 2]
```diff
@@ -26,14 +26,17 @@ ENV MODEL_NAME=model
 # Fix for the interactive mode during an install in step 21
```
Comment (arjkesh) [inline tensorflow/inference/docker/2.5/py3/cu112/Dockerfile.diy.gpu:26]: this file is from a poc and can be removed/hidden

File: tensorflow/training/docker/2.5/py3/Dockerfile.cpu [hunk 3]
```diff
@@ -140,7 +140,7 @@ RUN ln -s $(which ${PYTHON}) /usr/local/bin/python \
  && ln -s $(which ${PIP}) /usr/bin/pip
 
 RUN apt-get update && apt-get -y install protobuf-compiler \
-    && rm -rf /var/lib/apt/lists/* \
```
Comment (arjkesh) [inline tensorflow/training/docker/2.5/py3/Dockerfile.cpu:143]: I believe we need to update to 2.5.3 for TF - we may need a new binary as result

File: tensorflow/training/docker/2.5/py3/cu112/Dockerfile.gpu [hunk 4]
```diff
@@ -211,26 +211,29 @@ RUN ln -s $(which ${PYTHON}) /usr/local/bin/python \
 # install PyYAML==5.1.2 to avoid conflict with latest awscli
 # # python-dateutil==2.8.0 to satisfy botocore associated with latest awscli
 RUN ${PIP} install --no-cache-dir -U \
+   # Protobuf above this version appears to result in errors in horovod.
```
Comment (arjkesh) [inline tensorflow/training/docker/2.5/py3/cu112/Dockerfile.gpu:214]: I believe we need TF 2.5.3 in this container

File: tensorflow/inference/docker/2.5/py3/cu112/Dockerfile.gpu [hunk 5]
```diff
@@ -91,10 +94,14 @@ RUN ${PIP} --no-cache-dir install --upgrade \
 # We don't install libnvinfer-dev since we don't need to build against TensorRT
 # cuda-x.x is CUDA version 10.2, 11.0, or 11.1 (for 11.2)
 # https://docs.nvidia.com/deeplearning/tensorrt/archives/tensorrt-722/install-guide/index.html
-RUN apt-get update \
- && apt-get install -y --no-install-recommends --allow-unauthenticated  \
-    libnvinfer7=${TF_TENSORRT_VERSION}-1+cuda11.1 \
-    libnvinfer-plugin7=${TF_TENSORRT_VERSION}-1+cuda11.1 \
+# https://forums.developer.nvidia.com/t/notice-cuda-linux-repository-key-rotation/212771
```
Comment (arjkesh) [inline tensorflow/inference/docker/2.5/py3/cu112/Dockerfile.gpu:97]: can we use the latest version of tf serving api and model server? 2.5.4 I believe - https://pypi.org/project/tensorflow-serving-api/2.5.4/

File: tensorflow/training/docker/2.5/py3/cu112/Dockerfile.diy.gpu [hunk 6]
```diff
@@ -218,6 +218,7 @@ RUN ${PIP} install --no-cache-dir -U \
    # the library version to be overwritten
```
Comment (arjkesh) [inline tensorflow/training/docker/2.5/py3/cu112/Dockerfile.diy.gpu:218]: this is a POC file, we can remove/hide

**Review by arjkesh — COMMENTED** [review]
Do we also need to update TF inference 2.5?

File: tensorflow/training/docker/2.5/py3/Dockerfile.cpu [hunk 7]
```diff
@@ -118,29 +117,13 @@ RUN ${PIP} --no-cache-dir install --upgrade \
     pip \
     setuptools
```
Comment (tejaschumbalkar) [inline tensorflow/training/docker/2.5/py3/Dockerfile.cpu:119]: removed openssl installation from source in favor of apt installation

File: test/test_utils/__init__.py [hunk 8]
```diff
@@ -40,7 +40,7 @@ def get_ami_id_boto3(region_name, ami_name_pattern):
     return ami['ImageId']
 
     
-UBUNTU_18_BASE_DLAMI_US_WEST_2 = get_ami_id_boto3(region_name="us-west-2", ami_name_pattern="Deep Learning AMI GPU CUDA 11.1.1 (Ubuntu 18.04) ????????")
+UBUNTU_18_BASE_DLAMI_US_WEST_2 = "ami-0150e36b3f936a26e"
```
Comment (tejaschumbalkar) [inline test/test_utils/__init__.py:43]: reverted to old ami since the latest ami has some issues

File: test/test_utils/sagemaker.py [hunk 9]
```diff
@@ -53,7 +53,7 @@ def assign_sagemaker_remote_job_instance_type(image):
 
 def assign_sagemaker_local_job_instance_type(image):
     if "tensorflow" in image and "inference" in image and "gpu" in image:
-        return "p2.xlarge"
+        return "g4dn.xlarge"
```
Comment (tejaschumbalkar) [inline test/test_utils/sagemaker.py:56]: Changes the instance type for SM local test

**Comment by arjkesh** [comment]
labels changes LGTM - don't have any comments outside of the open ones

File: tensorflow/inference/docker/2.5/py3/cu112/Dockerfile.gpu (hunk 5 above)
Comment (tejaschumbalkar) [reply tensorflow/inference/docker/2.5/py3/cu112/Dockerfile.gpu:97]: Using the TFS2.5.1 as the rest of the patch releases are dummy

File: tensorflow/training/docker/2.5/py3/cu112/Dockerfile.gpu (hunk 4 above)
Comment (tejaschumbalkar) [reply tensorflow/training/docker/2.5/py3/cu112/Dockerfile.gpu:214]: Upgraded the patch version for training to 2.5.3

**Review by mseth10 — APPROVED** [review]
LGTM

File: tensorflow/training/docker/2.5/py3/Dockerfile.cpu (hunk 7 above)
Comment (saimidu) [reply tensorflow/training/docker/2.5/py3/Dockerfile.cpu:119]: I don't see the installation of openssl from apt in the dockerfile.

File: test/dlc_tests/sanity/test_pre_release.py [hunk 10]
```diff
@@ -412,9 +412,7 @@ def _run_dependency_check_test(image, ec2_connection):
         },
         "tensorflow": {
             "1.15": ["neuron"],
-            "2.5": ["neuron"],
-            "2.6": ["cpu", "gpu"],
-            "2.7": ["cpu", "gpu", "hpu"],
+            "2.5": ["cpu", "gpu", "neuron"],
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_pre_release.py:415]: Please correct this by adding the `2.6` and `2.7` key-value pairs back.

**Review by saimidu — COMMENTED** [review]
Mostly LGTM.

File: test/dlc_tests/sanity/test_pre_release.py (hunk 10 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/sanity/test_pre_release.py]: not sure on the origin of the change, added it back

**Review by saimidu — DISMISSED** [review]

**Review by saimidu — APPROVED** [review]
